### ISSUE TITLE
Add support for optional fields encoding masks

### ISSUE THREAD

**Issue by JulienBedel** [Orange-Cyberdefense/bof#12]
### TL;DR

In OPC UA, optional field presence is sometimes defined using an encoding mask. In that case, each digit of the mask represents if an optional field is present (1) or not (0).

This mechanism is not included in BOF yet, making it hard to represent frames with encoding mask based optional fields.

### Description

**Context**

I encountered this problem while implementing OPC UA protocol into BOF.

**Problem**

In OPC UA, optional field presence is sometimes defined using an encoding mask. In that case, each digit of the mask represents if an optional field is present (1) or not (0).

This mechanism is not included in BOF yet, making it hard to represent frames with encoding mask-based optional fields.

See *OPC 10000-6 - UA Specification Part 6 - Mappings 1.04, 5.2.7* for more information on the encoding mask.

**Example 1**

Let's take an `encoding_mask` field with value of `0x27`. It is `0010 1111` in binary, with each digit associated to the presence (or not) of an other field.
```
	0 : not used
	0 : has_symbolic_id
	1 : has_namespace
	0 : has_localized_text
	0 : has_locale
	1 : has_additional_info
	1 : has_inner_status_code
	1 : has_inner_diagnostic_info
```

### Solutions to investigate

**1. Naive solution**

The simplest solution I found for now is to use BOF current dependency system to associate every state of the encoding mask to the correspondent fields, but it implies to redefine every block combination which implies A LOT of duplication.

Let's imagine we have the following block :

```json
"TEST_BLOCK": [
    {"name": "encoding_mask", "type": "field", "size": 2},
    {"name": "encoded_block", "type": "depends:encoding_mask"},
],
```

We set up our codes like so :

```json
"node_id_value": {
        "00": "BLOCK_00000000",
        "01": "BLOCK_00000001",
        "02": "BLOCK_00000010",
        "03": "BLOCK_00000011"
        ...
}
```

And we have to define every block `BLOCK_XXXXXXXX` with or wihtout the fields we want in it or not.

This solution theoretically works, but with Example 1 above we have 2^8 = 256 combination possible. Which means 256 association to make in codes and 256 blocks to create..

**2. Multiple code associations**

Based on the same idea as the last solution, we could limit block duplication by allowing multiple code associations :

```json
"node_id_value": {
        "00": [],
        "01": ["DIAGNOSTIC_INFO"],
        "02": ["STATUS_CODE"],
        "03": ["DIAGNOSTIC_INFO", "STATUS_CODE"]
        ...
}
```
Therefore, we still have 255 code associations to make, but only 8 blocks to define.

**3. Bit field dependency**

```json
"TEST_BLOCK": [
    {"name": "encoding_mask", "type": "field", "size": 2},
    {"name": "has_symbolic_id", "type": "depends:encoding_mask[1]"},
    {"name": "has_namespace", "type": "depends:encoding_mask[2]"},
    {"name": "has_localized_text", "type": "depends:encoding_mask[3]"},
    {"name": "has_locale", "type": "depends:encoding_mask[4]"},
    {"name": "has_additional_info", "type": "depends:encoding_mask[5]"},
    {"name": "has_inner_status_code", "type": "depends:encoding_mask[6]"},
    {"name": "has_inner_diagnostic_info", "type": "depends:encoding_mask[7]"},
],`
```

This solution might be achieved based on existing BOF bitfields (but would still require to extend dependency search to bitfields) :
```json
"TEST_BLOCK": [
    {"name": "symbolic_id_encoding, namespace_encoding, .....", "type": "field", "size": 2, "bitsizes": "1, 1, ...."}
    {"name": "has_symbolic_id", "type": "depends:symbolic_id_encoding"},
    {"name": "has_namespace", "type": "depends:namespace_encoding"},
    ...
]
```

This solution may require only 8 block association (maybe none), depending on how it is implemented. It has the advantage of needing only 8 blocks definition.

### Help needed

Any help is appreciated, more precisely regarding the following topics :

- Identify the appropriate solution (⇒ trade-off with the ease of fix to think about).
- Help fixing the issue.

**Comment by claire-lex** [comment]
Thanks for submitting this issue.

I encountered a similar issue in KNX. I would suggest another workaround, close to your solution 3, the difference being that each bit field is referred to by its name.

In KNX, I also have to work with "masks", I handled them in the JSON specification file using ``bitfields``:

```
 {"name": "frame type, empty, repeat, broadcast type, priority, ack request, confirm", "type": "field", "size": 1, "bitsizes": "1, 1, 1, 1, 2, 1, 1"},
```

Obviously, it is not the nicest line in the file, but considering that BOF is designed to work with byte and that bit management is an add-on, it works!

Bitfields are "sub-fields" in a field, and a end-user should not have to know if the field she is changing in a script is a field or a bitfield: `frame_type.value = 1` where frame_type is a bitfield is equivalent to `my_field.value = 1` where my_field is a field. The conversion from int to byte or to bit is handled by BOF. 

This should also apply to the JSON: we should not be able to make a difference between a `depends:my_field` and a `depends:frame_type`. That's why I would recommend adding the support of bitfields when handling the ``depends`` keyword (``BOFBlock`` class in ``frame.py``).

Would that solve your problem?

**Comment by JulienBedel** [comment]
Thanks for the help @claire-lex, always appreciated !

In fact, your example would perfectly suits the needs of OPC UA masks too. We eventually reached the same solution of using already existing `bitfields` to handle the mask mechanisms. It now seems to be the right one !

I will try to add the support for `bitfields` for the `depends` keyword, and let you know if everything works as expected.

**Comment by JulienBedel** [comment]
I completed the `get_depends_value` to support bitfields and wrote a test accordingly.

@claire-lex, you can use the test as an example to try if it would fit the needs of KNX masks as well. I close the issue for now but feel free to tell me if you encounter a problem.

PS : while fixing the issue, I also had to slightly change specification handling to allow user-supplied json in blocks (to write tests). It is implemented in opcua's block _init_ and factory, feel free to get à look and see if it could be used for KNX too :)
